src/traceme.c

```diff
@@ -277,49 +321,103 @@ tracer_thread(void *arg)
-        tr_ctx->target_pid, -1, -1, 0);
-    if (tr_ctx->perf_fd == -1) {
-        err(EXIT_FAILURE, "syscall");
+    ret = syscall(SYS_perf_event_open, &attr, target_pid, -1, -1, 0);
+
+clean:
+    if (fclose(pt_type_file) == -1) {
+        ret = -1;
     }
 
+    return ret;
+}
+
+/*
+ * Set up Intel PT buffers and start a poll() loop for reading out the trace.
+ *
+ * Thread exits with 0 on success and -1 on error.
+ */
+static void *
+tracer_thread(void *arg)
+{
+    DEBUG("tracer init");
+
+    int page_size = getpagesize();
+    struct tracer_thread_args *thr_args = (struct tracer_thread_args *) arg;
+    int base_size = (1 + thr_args->data_bufsize) * page_size;
+    int ret = 0, sem_posted = 0;
+    void *base = NULL;
+    void *aux = NULL;
+    struct perf_event_mmap_page *header = NULL;
+
     /*
      * Allocate buffers.
      *
      * Data buffer is preceded by one management page, hence `1 + data_bufsize'.
      */
-    int page_size = getpagesize();
-    void *base = mmap(NULL, (1 + tr_ctx->data_bufsize) * page_size, PROT_WRITE,
-        MAP_SHARED, tr_ctx->perf_fd, 0);
+    base = mmap(NULL, base_size, PROT_WRITE,
+        MAP_SHARED, thr_args->perf_fd, 0);
     if (base == MAP_FAILED) {
-        err(EXIT_FAILURE, "mmap");
+        ret = -1;
+        goto clean;
     }
 
-    struct perf_event_mmap_page *header = base;
+    header = base;
     void *data = base + header->data_offset;
     (void) data; // XXX We will need this in the future to detect packet loss.
     header->aux_offset = header->data_offset + header->data_size;
-    header->aux_size = tr_ctx->aux_bufsize * page_size;
-    tr_ctx->mmap_header = header;
+    header->aux_size = thr_args->aux_bufsize * page_size;
 
     // AUX mapped R/W so as to have a saturating ring buffer.
-    void *aux = mmap(NULL, header->aux_size, PROT_READ | PROT_WRITE,
-        MAP_SHARED, tr_ctx->perf_fd, header->aux_offset);
+    aux = mmap(NULL, header->aux_size, PROT_READ | PROT_WRITE,
+        MAP_SHARED, thr_args->perf_fd, header->aux_offset);
     if (aux == MAP_FAILED) {
-        err(EXIT_FAILURE, "mmap2");
+        ret = -1;
+        goto clean;
     }
-    tr_ctx->aux = aux;
 
-    // Resume the interpreter loop.
+    // Copy out arguments for the poll loop.
+    int perf_fd = thr_args->perf_fd;
+    int stop_fd_rd = thr_args->stop_fd_rd;
+    int out_fd = thr_args->out_fd;
+
+    // Resume the interpreter loop. `thr_args' now unusable.
     DEBUG("resume main thread");
-    sem_post(&tr_ctx->tracer_init_sem);
+    if (sem_post(thr_args->tracer_init_sem) != 0) {
+        ret = -1;
+        goto clean;
+    }
+    sem_posted = 1;
 
     // Start reading out of the aux buffer.
-    poll_loop(tr_ctx);
+    if (poll_loop(perf_fd, stop_fd_rd, out_fd, header, aux) == -1) {
+        ret = -1;
+        goto clean;
+    }
 
-    DEBUG("tracer thread exit");
-    pthread_exit(NULL);
+clean:
+    if (aux) {
+        if (munmap(aux, header->aux_size) == -1) {
+            ret = -1;
```

Comment:
What's the point of setting `ret` at this point?

Reply:
Because this code is executed if we succeed and we want failure to clean up to be an error if you ask me.

Reply:
But we don't use `ret` past this point AFAICS.

Reply:
We use it to decide what to return later.

https://github.com/softdevteam/traceme/blob/9b62089022573b92fbf4d1b4aac811c2f5b390e4/src/traceme.c#L414-L420

There was a reason I didn't return `(intptr_t *) ret`. Let me check.

Reply:
Ah yes:
```
warning: src/traceme.c:422:18: warning: cast to 'intptr_t *' (aka 'long *') from smaller integer type 'int' [-Wint-to-pointer-cast]
warning:     pthread_exit((intptr_t *) ret);
warning:                  ^
warning: 1 warning generated.
```

Reply:
Don't you want `pthread_exit((intptr_t *) &ret)`?

Reply:
Eh? That would make a dangling pointer. Remember, we are using the pointer value as an integer.

Reply:
We want to return `0` for success or `-1` for failure. We just happen to pack it into a void pointer as that's all the pthread interface gives us.

`&ret` would always indicate non-zero, as the address of ret won't ever be null.

Reply:
Ah, I see what's going on: it's the "`void *` is guaranteed to be able to hold any other value" trick. A bit evil, but kind of makes sense.

Reply:
It's either that or use a real pointer and find somewhere for what it points to to live.

It's a shortcoming of the pthread api due to C's type system really.

Reply:
Right, but the problem then becomes that we're *also* abusing `intptr_t` and pretending that 0 and 1 are addresses. Why don't we just be honest and cast an `int` to a `void *` somehow (avoiding the compiler warning)?

Reply:
I'm not sure there's a portable way. I mean, I could cast `int` to `long long` and then to `void *`, but this assumes x86_64.

As far as I can see, `intptr_t` is exactly what we want. It was introduced into C99 as a:
> [signed] integer type capable of holding a pointer

http://en.cppreference.com/w/c/types/integer

Reply:
Another dumb question: shouldn't we be using `intptr_t` and not `intptr_t *`?

Reply:
oh. good question.

Reply:
I don't think we are allowed to cast 0 and 1 to intptr_t:
```
warning: src/traceme.c:418:22: warning: expression which evaluates to zero treated as a null pointer constant of type 'void *' [-Wnon-literal-null-conversion]
warning:         pthread_exit((intptr_t) 0);
warning:                      ^~~~~~~~~~~~
warning: src/traceme.c:421:22: warning: incompatible integer to pointer conversion passing 'intptr_t' (aka 'long') to parameter of type 'void *' [-Wint-conversion]
warning:         pthread_exit((intptr_t) -1);
warning:                      ^~~~~~~~~~~~~
warning: /usr/include/pthread.h:242:33: note: passing argument to parameter '__retval' here
warning: extern void pthread_exit (void *__retval) __attribute__ ((__noreturn__));
warning:                                 ^
warning: 2 warnings generated.
```

Any other ideas? All the examples on the net cast to `void *`, which seems to work.

Let me think about this a bit more...

Reply:
I think we can actually kill the explicit calls to `pthread_exit`.  Negative numbers should be fine as long as you cast to the same type on the other side, but if you wanted bools, we could use a bool here too.

Reply:
I'm nervous about negative numbers because funny things might happen due to our not fully understanding the various casting rules. Best not to risk it IMHO.

If we don't need the calls, then let's kill them. If we need them, maybe we should just define constants like `THREAD_SUCCESS` / `THREAD_FAILURE` and so on? Then it doesn't matter what (positive) integer we set them to.

Reply:
You are over-thinking it :) stdbool is basically an unsigned int.

Reply:
Yes, but what if we want other return codes? Or will success/failure only ever be it?